### PR TITLE
DT: Add DT_DMA macros to help usage in dma clients drivers 

### PR DESCRIPTION
**Pull request by erwango**
Provide DT macros to help dma properties access in DMA clients.
Tests are provided as well.

Additionally, convert stm32 i2s driver to demonstrate the intended use.

This PR is the official version of https://github.com/zephyrproject-rtos/zephyr/pull/24123

EDIT:
Tested on x_nucleo_iks02a1 shield

### CODE REVIEW

File: include/devicetree/dma.h
Comment (avisconti) [inline include/devicetree/dma.h:7]: shouldn't be ST ?

**Comment by avisconti** [comment]
Where did you have try this out? Can you add it in the PR description?
Did you test it on shield iks02a1 as well?

File: include/devicetree/dma.h
Comment (erwango) [reply include/devicetree/dma.h:7]: No :-)

**Comment by avisconti** [comment]
> Where did you have try this out? Can you add it in the PR description?
> Did you test it on shield iks02a1 as well?

I tested it myself on iks02a1 shield. It worked fine.

**Comment by erwango** [comment]
> Where did you have try this out? Can you add it in the PR description?
Did you test it on shield iks02a1 as well?

I was missing hw to test. I was hopping you could help on that

**Comment by erwango** [comment]
> I tested it myself on iks02a1 shield. It worked fine.

Great, thanks!

File: include/devicetree/dma.h
Comment (erwango) [reply include/devicetree/dma.h:7]: No :-)

**Review by galak — APPROVED** [review]
LGTM.  Would like to get @mbolivar-nordic approval as well before merging.

**Review by avisconti — APPROVED** [review]

File: include/devicetree/dma.h
Comment (mbolivar-nordic) [inline include/devicetree/dma.h:239]:
```suggestion
```

File: include/devicetree/dma.h
Comment (mbolivar-nordic) [inline include/devicetree/dma.h:234]:
```suggestion
 * @brief Does a DMA client have a DMA specifier with a given name?
```

File: include/devicetree/dma.h
Comment (mbolivar-nordic) [inline include/devicetree/dma.h:249]:
```suggestion
```

File: include/devicetree/dma.h
Comment (mbolivar-nordic) [inline include/devicetree/dma.h:244]:
```suggestion
 * @brief Does a DMA client have a DMA specifier with a given name?
```

File: include/devicetree/dma.h
Comment (mbolivar-nordic) [inline include/devicetree/dma.h:261]:
```suggestion
```

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
Looks good overall!

Review summary:

- testing is not following API guarantees regarding instance numbers, with possible suggested fix. This is a blocker but should be easy to fix.
- various bikeshedding comments about documentation, done as GitHub suggestions you can click to apply if you choose.

File: include/devicetree/dma.h
Comment (erwango) [reply include/devicetree/dma.h:52]: Right. Actually this is due to a three steps implementation :-). I'll fix that.

File: tests/lib/devicetree/src/main.c
Comment (erwango) [reply tests/lib/devicetree/src/main.c:888]: ok, then, since I just reused "test_spi" node, do I really need to add an additional test? Or is it just superfluous?

File: include/devicetree/dma.h
Comment (erwango) [reply include/devicetree/dma.h:52]: I think it should be ok now

**Comment by erwango** [comment]
@mbolivar-nordic , thanks for the review and suggestions, I think I'm done.

File: tests/lib/devicetree/src/main.c
Comment (mbolivar-nordic) [reply tests/lib/devicetree/src/main.c]: Sorry I was vague. All of the new inst macros need test cases. However, the TEST_SPI node is not the right place to add the "dmas" property which is used to test these macros.

That is because there are two instances of "vnd,spi" in the devicetree. This means that hard-coding instance number 0, for example as done here:

```
	zassert_equal(DT_INST_DMAS_CELLS_BY_NAME(0, rx, channel), 3,
		      "channel cell of rx dma channel");
```

Is not correct, because there is no guarantee that `/test/spi@33334444` is instance number 0 of "vnd,spi". It could be `/test/spi@55556666`; the API does not make a guarantee.

The solution is to put the "dmas" and "dma-names" properties inside of a node whose compatible has exactly one instance in the entire devicetree. Then the API guarantee that the instances are numbered starting at 0 for every enabled node means that the above test would be safe.

So I am suggesting doing something like this instead:

```
		/* there should only be one of these */
		test_temp_sensor: temperature-sensor {
			compatible = "vnd,adc-temp-sensor";
			label = "TEST_TEMP";
			dmas = <&test_dma1 1 2>, <&test_dma2 3 4>;
			dma-names = "tx", "rx";
			io-channels = <&test_adc_1 10>, <&test_adc_2 20>;
			io-channel-names = "ch1", "ch2";
			clocks = <&test_clk 3 7>, <&test_fixed_clk>, <&test_clk 8 2>;
		};
```

And doing the tests with DT_DRV_COMPAT set to vnd_adc_temp_sensor.

You can pick any node that we ensure there's only one of. This is just a suggestion.

Comment (erwango) [reply tests/lib/devicetree/src/main.c]: > Is not correct, because there is no guarantee that /test/spi@33334444 is instance number 0 of "vnd,spi". It could be /test/spi@55556666; the API does not make a guarantee.

Ok. This was indeed what I was trying to check.
I get your point now, I'll take your suggestion.

Comment (erwango) [reply tests/lib/devicetree/src/main.c]: @mbolivar-nordic  Done

**Review by mbolivar-nordic — APPROVED** [review]
Thanks, tests look good now.

**Pull request merged**
